### PR TITLE
Feature: Offline donations gateway v3 compatibility

### PR DESCRIPTION
**Pull request by JasonTheAdams**
## Description

This PR is a rebuild of the Offline gateway to make use of the Gateway API and be v2 + v3 compatible. It cleans up a bunch of the old code and should port everything over nicely. The ID and everything should be preserved such that it's full backwards-compatible.

Note, since there's now a Billing Address block, the billing address settings don't do anything anymore on v3 forms. It should still work on v2 forms, though.

## Affects

The offline donations gateway

### In the donation form
image

### Settings in form builder
image

## Testing Instructions

Make sure the Offline gateway works with v2 and v3 forms. Make sure to test out the various settings and emails related to offline.

## Pre-review Checklist

-   [x] Relevant `@unreleased` tags included in DocBlocks
-   [x] Includes unit tests
-   [x] Reviewed by the designer (if follows a design)
-   [x] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

### CODE REVIEW

**Comment by JasonTheAdams** [comment]
@jonwaldstein This is ready for review! Happy to hop on a call to review together if you feel that would speed things up.

**Comment by jonwaldstein** [comment]
@JasonTheAdams i'm getting a bunch of errors when running `npm run dev`.  If you added a typescript package to support an existing package - the versions need to match up or else the types are coming from a different version.

Screenshot 2023-09-07 at 4 17 49 PM

**Comment by JasonTheAdams** [comment]
@jonwaldstein Are you able to help with that? How do I know which version to use?

**Comment by jonwaldstein** [comment]
> @jonwaldstein Are you able to help with that? How do I know which version to use?

The version of the `@types/{package}` and the actual `{package}` need to be as close as possible - since the `@types/{package}` is basically supplying types for that version of the package.   

Yes I can help

**Comment by jonwaldstein** [comment]
@JasonTheAdams fyi there's some discrepencies between the latest `@types/wordpress__block-editor` version and `@wordpress/block-editor` - they're not in sync at the moment so removing for now.  we can solve this another time.

**Review by jonwaldstein — CHANGES_REQUESTED** [review]
Looking good, left a few comments.  You will also want to add a migration for the form settings to block attributes.

File: src/PaymentGateways/Gateways/Offline/OfflineGateway.php [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+use Give\Framework\PaymentGateways\Commands\PaymentPending;
+use Give\Framework\PaymentGateways\PaymentGateway;
+use Give\Framework\Support\Facades\Scripts\ScriptAsset;
+use Give\Helpers\Language;
+use Give\PaymentGateways\Gateways\Offline\Views\LegacyFormFieldMarkup;
+
+/**
+ * The Offline payment gateway, intended to reflect donations that are made offline and will be later confirmed.
+ */
+class OfflineGateway extends PaymentGateway
```
[the remark this answers is not in the document]
Comment (JasonTheAdams) [reply src/PaymentGateways/Gateways/Offline/OfflineGateway.php:18]: Resolved in 37ddc7da6

File: src/PaymentGateways/Gateways/Offline/Actions/DisableGatewayWhenDisabledPerForm.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        }
+
+        $form = DonationForm::find($formId);
+
+        $paymentGatewaysBlock = $form->blocks->findByName('givewp/payment-gateways');
+        if (!$paymentGatewaysBlock) {
+            throw new RuntimeException('Payment gateways block not found');
+        }
+
+        ['offlineEnabled' => $enabled] = $paymentGatewaysBlock->getAttributes();
```
[the remark this answers is not in the document]
Comment (JasonTheAdams) [reply src/PaymentGateways/Gateways/Offline/Actions/DisableGatewayWhenDisabledPerForm.php:41]: Good catch. Resolved in f71b29857

File: src/FormBuilder/resources/js/form-builder/src/components/settings/PopoverContentWithTemplateTags/index.tsx [hunk 3]
```diff
@@ -58,6 +60,7 @@ export default function PopoverContentWithTemplateTags({
     onClose,
     templateTags,
     heading,
+    useEditor,
```
[the remark this answers is not in the document]
Comment (JasonTheAdams) [reply src/FormBuilder/resources/js/form-builder/src/components/settings/PopoverContentWithTemplateTags/index.tsx]: I think that's most relevant for functions. I don't think it matters for props. But not a big deal, resolved in 2bcebdeee

**Comment by jonwaldstein** [comment]
@JasonTheAdams are you going to update this PR with the migration as well?

**Comment by JasonTheAdams** [comment]
> @JasonTheAdams are you going to update this PR with the migration as well?

Ah yes, good point! I'll do that now!

**Comment by JasonTheAdams** [comment]
@jonwaldstein I added the migration code: 22fff1157

**Review by jonwaldstein — APPROVED** [review]
Nice job! ready for QA 💸

**Review by rickalday — APPROVED** [review]
Passed manual QA tests.

**Pull request merged**